runner/runner.go

```diff
@@ -0,0 +1,379 @@
+	}
+
+	if disabledSleep > 0 {
+		if debug {
+			logger.Printf("Automatic execution disabled, sleeping for an additional %d minutes", disabledSleepSeconds/60)
+		}
+
+		time.Sleep(disabledSleep)
+	} else if debug {
+		logger.Println("Automatic execution disabled")
+	}
+
+	return false
+}
+
+func getMultisiteSites() ([]site, error) {
+	raw, err := runWpCliCmd([]string{"site", "list", "--fields=url", "--archived=false", "--deleted=false", "--spam=false", "--format=json"})
+	if err != nil {
+		return make([]site, 0), err
+	}
+
+	jsonRes := make([]site, 0)
+	if err = json.Unmarshal([]byte(raw), &jsonRes); err != nil {
+		if debug {
+			logger.Println(fmt.Sprintf("%+v\n", err))
+		}
+
+		return make([]site, 0), err
+	}
+
+	// Shuffle site order so that none are favored
+	for i := range jsonRes {
+		j := rand.Intn(i + 1)
+		jsonRes[i], jsonRes[j] = jsonRes[j], jsonRes[i]
+	}
+
+	return jsonRes, nil
+}
+
+func queueSiteEvents(workerID int, sites <-chan site, queue chan<- event) {
+	for site := range sites {
+		if debug {
+			logger.Printf("getEvents-%d processing %s", workerID, site.URL)
+		}
+
+		events, err := getSiteEvents(site.URL)
+		if err != nil {
+			time.Sleep(getEventsBreak)
+			continue
+		}
+
+		for _, event := range events {
+			event.URL = site.URL
+			queue <- event
+		}
+
+		time.Sleep(getEventsBreak)
+	}
+}
+
+func getSiteEvents(site string) ([]event, error) {
+	raw, err := runWpCliCmd([]string{"cron-control", "orchestrate", "runner-only", "list-due-batch", fmt.Sprintf("--url=%s", site), "--format=json"})
+	if err != nil {
+		return make([]event, 0), err
+	}
+
+	siteEvents := make([]event, 0)
+	if err = json.Unmarshal([]byte(raw), &siteEvents); err != nil {
+		if debug {
+			logger.Println(fmt.Sprintf("%+v\n", err))
+		}
+
+		return make([]event, 0), err
+	}
+
+	return siteEvents, nil
+}
+
+func runEvents(workerID int, events <-chan event) {
+	for event := range events {
+		if now := time.Now(); event.Timestamp > int(now.Unix()) {
+			if debug {
+				logger.Printf("runEvents-%d skipping premature job %d|%s|%s for %s", workerID, event.Timestamp, event.Action, event.Instance, event.URL)
+			}
+
+			continue
+		}
+
+		subcommand := []string{"cron-control", "orchestrate", "runner-only", "run", fmt.Sprintf("--timestamp=%d", event.Timestamp), fmt.Sprintf("--action=%s", event.Action), fmt.Sprintf("--instance=%s", event.Instance), fmt.Sprintf("--url=%s", event.URL)}
+
+		_, err := runWpCliCmd(subcommand)
+
+		if err == nil {
+			if heartbeatInt > 0 {
+				atomic.AddUint64(&eventRunSuccessCount, 1)
+			}
+
+			if debug {
+				logger.Printf("runEvents-%d finished job %d|%s|%s for %s", workerID, event.Timestamp, event.Action, event.Instance, event.URL)
+			}
+		} else if heartbeatInt > 0 {
+			atomic.AddUint64(&eventRunErrCount, 1)
+		}
+
+		time.Sleep(runEventsBreak)
+	}
+}
+
+func runWpCliCmd(subcommand []string) (string, error) {
+	subcommand = append(subcommand, "--allow-root", "--quiet", fmt.Sprintf("--path=%s", wpPath))
```

Comment:
It's necessary anytime we request data in JSON format, as it suppresses the "informational" stuff that commands include. Without it, we often don't get a string that can be unmarshalled.